drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+
+	data->scan_rslt_cb = cb;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	ret = whd_wifi_scan(airoc_if, params->scan_type, WHD_BSS_TYPE_ANY, &(data->ssid), NULL,
+			    NULL, NULL, scan_callback, &(data->scan_result), data);
+
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int airoc_mgmt_connect(const struct device *dev, struct wifi_connect_req_params *params)
+{
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+	whd_ssid_t ssid = {0};
+	int ret = 0;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (data->is_sta_connected) {
+		LOG_ERR("Already connected");
+		ret = -EALREADY;
+		goto error;
+	}
+
+	ssid.length = params->ssid_length;
+	memcpy(ssid.value, params->ssid, params->ssid_length);
+
+	whd_scan_result_t scan_result;
+	whd_scan_result_t usr_result = {0};
+
+	usr_result.SSID.length = ssid.length;
+	memcpy(usr_result.SSID.value, ssid.value, ssid.length);
+
+	if (whd_wifi_scan(airoc_if, WHD_SCAN_TYPE_ACTIVE, WHD_BSS_TYPE_ANY, NULL, NULL, NULL, NULL,
+			  scan_cb_search, &scan_result, &(usr_result)) != WHD_SUCCESS) {
+		LOG_ERR("Failed start scan");
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (k_sem_take(&airoc_wifi_data.sema_scan, K_MSEC(12 * 1000)) != 0) {
+		whd_wifi_stop_scan(airoc_if);
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (usr_result.security == 0) {
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	/* Connect to the network */
+	if (whd_wifi_join(airoc_if, &usr_result.SSID, usr_result.security, params->psk,
+			  params->psk_length) != WHD_SUCCESS) {
+		LOG_ERR("Failed to connect with network");
+		ret = -EAGAIN;
+		goto error;
+	}
+
+error:
+	if (ret < 0) {
+		net_if_dormant_on(data->iface);
+	} else {
+		net_dhcpv4_start(data->iface);
+		net_if_dormant_off(data->iface);
+		data->is_sta_connected = 1;
+	}
+
+	wifi_mgmt_raise_connect_result_event(data->iface, ret);
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int airoc_mgmt_disconnect(const struct device *dev)
+{
+	int ret = 0;
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (whd_wifi_leave(airoc_if) != WHD_SUCCESS) {
+		k_sem_give(&data->sema_common);
+		ret = -EAGAIN;
+	} else {
+		data->is_sta_connected = 0;
+		net_dhcpv4_stop(data->iface);
+		net_if_dormant_on(data->iface);
+	}
+
+	wifi_mgmt_raise_disconnect_result_event(data->iface, ret);
+	k_sem_give(&data->sema_common);
+
+	return ret;
+}
+
+static void *ap_link_events_handler(whd_interface_t ifp, const whd_event_header_t *event_header,
+				    const uint8_t *event_data, void *handler_user_data)
+{
+	struct airoc_wifi_event_t airoc_event = {
+		.is_ap_event = 1,
+		.event_type = event_header->event_type
+	};
+
+	k_msgq_put(&airoc_wifi_msgq, &airoc_event, K_FOREVER);
```

Comment:
k_msgq_put returns a value. Should we assert if its not success? This function doesn't return a status, so that seems the reasonable thing to do.

Reply:
i do not think, that we need to add something here. k_msgq_put call with K_FOREVER, so it will wait until  put the msg..